File: src/app/src/main/java/com/huawei/myhealth/java/ui/fragments/SignInFragment.java
/*
 *
 *  * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *  *
 *  *    Licensed under the Apache License, Version 2.0 (the "License");
 *  *    you may not use this file except in compliance with the License.
 *  *    You may obtain a copy of the License at
 *  *
 *  *      http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  *    Unless required by applicable law or agreed to in writing, software
 *  *    distributed under the License is distributed on an "AS IS" BASIS,
 *  *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *    See the License for the specific language governing permissions and
 *  *    limitations under the License.
 *
 */
package com.huawei.myhealth.java.ui.fragments;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Paint;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;

import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.material.snackbar.Snackbar;
import com.huawei.agconnect.auth.AGConnectAuth;
import com.huawei.agconnect.auth.AGConnectAuthCredential;
import com.huawei.agconnect.auth.AGConnectUser;
import com.huawei.agconnect.auth.PhoneAuthProvider;
import com.huawei.agconnect.auth.SignInResult;
import com.huawei.agconnect.auth.VerifyCodeSettings;
import com.huawei.hmf.tasks.OnFailureListener;
import com.huawei.hmf.tasks.OnSuccessListener;
import com.huawei.hmf.tasks.Task;
import com.huawei.myhealth.R;
import com.huawei.myhealth.databinding.FragmentSigninBinding;
import com.huawei.myhealth.java.database.dbabstract.MyHealthDatabase;
import com.huawei.myhealth.java.database.entity.UserProfile;
import com.huawei.myhealth.java.utils.CommonUtil;
import com.huawei.myhealth.java.utils.DefaultValues;

import java.util.List;
import java.util.Locale;
/**
 * @since 2020
 * @author Huawei DTSE India
 */

public class SignInFragment extends Fragment implements View.OnClickListener {
    private String TAG = "SignInFragment";
    MyHealthDatabase db;
    FragmentSigninBinding binding;
    boolean mobileexist = false;



    public SignInFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentSigninBinding.inflate(inflater, container, false);
        //db initialization
        db = MyHealthDatabase.getDatabase(getActivity());

        // To Show / Hide Action Bar
        AppCompatActivity activity = (AppCompatActivity) getActivity();
        if (activity != null) {
            activity.getSupportActionBar().show();
        }

        // set underline code for resend text
        binding.tvResendcode.setPaintFlags( binding.tvResendcode.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);

        // check for signIn
        isAlreadySignedIn();

        // signIn button action performed
        binding.signInBtn.setOnClickListener(this);

        // mobile edit text listner
        binding.enterMobileNoET.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                if (editable.toString().length() < 10) {
                    changeBtnText(getResources().getString(R.string.sign_in));
                    binding.enterOTPTIL.setVisibility(View.GONE);
                    binding.tvResendcode.setVisibility(View.GONE);
                }
            }
        });

        // resend code action performed
        binding.tvResendcode.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String mobileNoStr = binding.enterMobileNoET.getText().toString();
                applyAndVerifyMobileNo(mobileNoStr);
            }
        });

        return binding.getRoot();

    }

    @Override
    public void onClick(View view) {
        String mobileNoStr = binding.enterMobileNoET.getText().toString();
        if (binding.signInBtn.getText().toString().equals(getResources().getString(R.string.sign_in))) {
            if ((!mobileNoStr.isEmpty() )&& mobileNoStr.length() == 10) {
                applyAndVerifyMobileNo(mobileNoStr);
            } else {
                Snackbar.make(binding.signInParentLay, getResources().getString(R.string.enter_validnumber), Snackbar.LENGTH_LONG).show();
            }
        } else {
            String otpStr = binding.enterOTPET.getText().toString();
            verifyOTP(mobileNoStr, otpStr);
        }
    }

    private void applyAndVerifyMobileNo(String mobileNoStr)
    {
        showHideProgressbar(true);
        VerifyCodeSettings settings = VerifyCodeSettings.newBuilder().action(1001).sendInterval(DefaultValues.VALUE_30).locale(Locale.ENGLISH).build();
        final Task task = PhoneAuthProvider.requestVerifyCode(getResources().getString(R.string.mobilecode), mobileNoStr, settings);
        task.addOnSuccessListener(new OnSuccessListener() {
                                      @Override
                                      public void onSuccess(Object o) {
                                          // onSuccess
                                          task.isSuccessful();
                                          showHideProgressbar(false);
                                          binding.enterOTPTIL.setVisibility( View.VISIBLE);
                                          binding.tvResendcode.setVisibility(View.VISIBLE);
                                          changeBtnText(getResources().getString(R.string.submit_OTP));
                                          Snackbar.make(binding.signInParentLay, getResources().getString(R.string.otp_sent), Snackbar.LENGTH_LONG).show();
                                      }
                                  }
        ).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(Exception e) {
                // onFail
                Snackbar.make(binding.signInParentLay, e.toString(), Snackbar.LENGTH_LONG).show();
                showHideProgressbar(false);
                // binding.enterOTPTIL.visibility = View.GONE
                changeBtnText(getResources().getString(R.string.sign_In));
            }
        });
    }

    private void verifyOTP(String mobileNoStr,String otpStr)
    {
        AGConnectAuthCredential credential = PhoneAuthProvider.credentialWithVerifyCode(getResources().getString(R.string.mobilecode), mobileNoStr, null, otpStr);
        AGConnectAuth.getInstance().signIn(credential).addOnSuccessListener(new OnSuccessListener<SignInResult>() {
            @Override
            public void onSuccess(SignInResult signInResult) {
                // Obtain sign-in information.
                mobileexist = false;
                Thread thread = new Thread(new Runnable() {
                    @Override
                    public void run() {
                        // fetch Records
                        List<UserProfile> list_userprofile = db.userProfileDao().getUserData();
                        for(int i=0; i<list_userprofile.size(); i++) {
                            if(binding.enterMobileNoET.getText().toString().equals(list_userprofile.get(0).getMobileNoStr()))
                            {
                                mobileexist = true;
                            }
                        }
                    }
                });

                thread.start();
                thread.interrupt();

                if (mobileexist) {
                    NavHostFragment.findNavController(getParentFragment()).navigate(R.id.homeScreenDest);
                } else {
                    NavHostFragment.findNavController(getParentFragment()).navigate(R.id.completeProfileFragmentDest);
                }
                // storing in shared preference
                SharedPreferences sharedPreference = getActivity().getSharedPreferences(getResources().getString(R.string.APP), Context.MODE_PRIVATE);
                SharedPreferences.Editor editor = sharedPreference.edit();
                editor.putString(getResources().getString(R.string.mobilenum), binding.enterMobileNoET.getText().toString());
                editor.commit();
                binding.enterMobileNoET.setText("");
                binding.enterOTPET.setText("");
            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(Exception e) {
                CommonUtil.logger(e.toString(),TAG);
                if (e.getMessage().contains(getResources().getString(R.string.already_user))) {
                    NavHostFragment.findNavController(getParentFragment()).navigate(R.id.homeScreenDest);
                } else {
                    isProfileCompleted();
                }
            }
        });
    }

    private void changeBtnText(String textOnBtn) {
        binding.signInBtn.setText(textOnBtn);
    }

    private void showHideProgressbar(boolean isVisibleCustom) {
        if(isVisibleCustom) {
            binding.signInProgressbar.setVisibility(View.VISIBLE);
        }
        else
        {
            binding.signInProgressbar.setVisibility(View.GONE);
        }
    }

    private void isAlreadySignedIn() {
        AGConnectUser user = AGConnectAuth.getInstance().getCurrentUser();
        if (user != null) {
            isProfileCompleted();
        }
    }

    private void isProfileCompleted() {
        AGConnectUser user = AGConnectAuth.getInstance().getCurrentUser();
        if (user.getDisplayName() == null) {
            //do nothing
        } else {
            NavHostFragment.findNavController(getParentFragment()).navigate(R.id.homeScreenDest);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: src/app/src/main/java/com/huawei/myhealth/kotlin/ui/fragments/SignInFragment.kt
/*
 *
 *  * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *  *
 *  *    Licensed under the Apache License, Version 2.0 (the "License");
 *  *    you may not use this file except in compliance with the License.
 *  *    You may obtain a copy of the License at
 *  *
 *  *      http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  *    Unless required by applicable law or agreed to in writing, software
 *  *    distributed under the License is distributed on an "AS IS" BASIS,
 *  *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *    See the License for the specific language governing permissions and
 *  *    limitations under the License.
 *
 */

package com.huawei.myhealth.ui.fragments

import android.content.Context
import android.graphics.Paint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import androidx.core.widget.doAfterTextChanged
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import androidx.room.Room
import com.google.android.material.snackbar.Snackbar
import com.huawei.agconnect.auth.AGConnectAuth
import com.huawei.agconnect.auth.PhoneAuthProvider
import com.huawei.agconnect.auth.VerifyCodeSettings
import com.huawei.myhealth.R
import com.huawei.myhealth.database.dbabstract.MyHealthDatabase
import com.huawei.myhealth.databinding.FragmentSigninBinding
import com.huawei.myhealth.utils.CommonUtil
import com.huawei.myhealth.utils.DefaultValues
import java.util.*
/**
 * @since 2020
 * @author Huawei DTSE India
 */

class SignInFragment : Fragment(), View.OnClickListener {
    private val TAG = "SignInFragment"
    var db: MyHealthDatabase? = null
    private var _binding: FragmentSigninBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentSigninBinding.inflate(inflater, container, false)
        val view = binding.root

        // db initilize
        db = activity?.let { Room.databaseBuilder(it, MyHealthDatabase::class.java, resources.getString(R.string.my_health_database)).build() }

        // To Show / Hide Action Bar
        (activity as AppCompatActivity?)!!.supportActionBar!!.show()

        // set underline code for resend text
        binding.tvResendcode.setPaintFlags(binding.tvResendcode.getPaintFlags() or Paint.UNDERLINE_TEXT_FLAG)

        // check for signIn
        isAlreadySignedIn()

        // signIn button action performed
        binding.signInBtn.setOnClickListener(this)

        // mobile edit text listner
        binding.enterMobileNoET.doAfterTextChanged {
            if (binding.enterMobileNoET.text?.length!! < 10) {
                changeBtnText(resources.getString(R.string.sign_in))
                binding.enterOTPTIL.visibility = View.GONE
                binding.tvResendcode.visibility = View.GONE
            }
        }

        // resend code action performed
        binding.tvResendcode!!.setOnClickListener {
            val mobileNoStr: String = binding.enterMobileNoET.text.toString()
            applyAndVerifyMobileNo(mobileNoStr)
        }
        return view
    }
    private fun applyAndVerifyMobileNo(mobileNoStr: String) {

            showHideProgressbar(true)
            val settings = VerifyCodeSettings.newBuilder()
                    .action(VerifyCodeSettings.ACTION_REGISTER_LOGIN)
                    .sendInterval(DefaultValues.VALUE_30)
                    .locale(Locale.ENGLISH)
                    .build()
            val task = PhoneAuthProvider.requestVerifyCode(resources.getString(R.string.mobilecode), mobileNoStr, settings)
            task.addOnSuccessListener {
                // onSuccess
                task.isSuccessful
                showHideProgressbar(false)
                binding.enterOTPTIL.visibility = View.VISIBLE
                binding.tvResendcode.visibility = View.VISIBLE
                changeBtnText(resources.getString(R.string.submit_OTP))
                Snackbar.make(binding.signInParentLay, resources.getString(R.string.otp_sent), Snackbar.LENGTH_LONG).show()
            }.addOnFailureListener { e ->
                // onFail
                Snackbar.make(binding.signInParentLay, e.toString(), Snackbar.LENGTH_LONG).show()
                showHideProgressbar(false)
                // binding.enterOTPTIL.visibility = View.GONE
                changeBtnText(resources.getString(R.string.sign_In))
            }

    }

    private fun verifyOTP(mobileNoStr: String, otpStr: String) {
        val credential = PhoneAuthProvider.credentialWithVerifyCode(resources.getString(R.string.mobilecode), mobileNoStr, null, otpStr)

        AGConnectAuth.getInstance().signIn(credential)
                .addOnSuccessListener {
                    // Obtain sign-in information.
                    var mobileexist = false
                    //fetch Records
                    val thread = Thread {
                        db?.userProfileDao()?.getUserData()?.forEach()
                        {
                            if (binding.enterMobileNoET.text.toString().equals(it.mobileNoStr)) {
                                mobileexist = true
                            }
                        }
                    }
                    thread.start()
                    thread.interrupt()
                    if (mobileexist) {
                        findNavController().navigate(R.id.homeScreenDest)
                    } else {
                        findNavController().navigate(R.id.completeProfileFragmentDest)
                    }
                    // storing in shared preference
                    val sharedPreference = activity?.getSharedPreferences(resources.getString(R.string.APP), Context.MODE_PRIVATE)
                    var editor = sharedPreference?.edit()
                    editor?.putString(resources.getString(R.string.mobilenum), binding.enterMobileNoET.text.toString())
                    editor?.commit()
                    binding.enterMobileNoET.setText("")
                    binding.enterOTPET.setText("")
                }
                .addOnFailureListener { e ->
                    CommonUtil.logger(e.toString(),TAG)
                    if (e.message?.contains(resources.getString(R.string.already_user))!!) {
                        findNavController().navigate(R.id.homeScreenDest)
                    } else {
                        isProfileCompleted()
                    }
                }
    }

    private fun changeBtnText(textOnBtn: String) {
        binding.signInBtn.text = textOnBtn
    }

    private fun showHideProgressbar(isVisibleCustom: Boolean) {
        binding.signInProgressbar.isVisible = isVisibleCustom
    }

    private fun isAlreadySignedIn() {
        val user = AGConnectAuth.getInstance().currentUser
        if (user != null) {
            isProfileCompleted()
        }
    }

    private fun isProfileCompleted() {
        val user = AGConnectAuth.getInstance().currentUser
        if (user.displayName == null) {
            //do nothing
        } else {
            findNavController().navigate(R.id.homeScreenDest)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()

        _binding = null
    }

    override fun onClick(v: View?) {
        val mobileNoStr: String = binding.enterMobileNoET.text.toString()
        if (binding.signInBtn.text.toString() == resources.getString(R.string.sign_in)) {
            if (mobileNoStr.isNotBlank() && mobileNoStr.length == 10) {

                applyAndVerifyMobileNo(mobileNoStr)
            } else {
                Snackbar.make(binding.signInParentLay, resources.getString(R.string.enter_validnumber), Snackbar.LENGTH_LONG).show()
            }
        } else {
            val otpStr: String = binding.enterOTPET.text.toString()

            verifyOTP(mobileNoStr, otpStr)
        }
    }
}